app/gui2/shared/ast/parse.ts

```diff
@@ -591,11 +596,128 @@ function resync(
     const goodAst = goodSpans.get(span)?.[0]
     // The parent of the root of a bad subtree must be a good AST.
     assertDefined(goodAst)
-    parent.replaceValue(edit.copy(goodAst))
+    parent.syncToCode(goodAst.code())
   }
 
   console.warn(
     `repair: Replaced ${parentsOfBadSubtrees.size} subtrees with their reparsed equivalents.`,
     parentsOfBadSubtrees,
   )
 }
+
+function hashString(s: string) {
+  return hash(s)
+}
+
+function hashSubtree(ast: Ast, hashesOut: Map<string, Ast[]>) {
+  let content = ''
+  content += ast.typeName + ':'
+  for (const child of ast.concreteChildren()) {
+    content += child.whitespace ?? ' '
+    if (isTokenId(child.node)) {
+      content += 'Token:' + hashString(ast.module.getToken(child.node).code())
+    } else {
+      content += hashSubtree(ast.module.checkedGet(child.node), hashesOut)
+    }
+  }
+  const astHash = hashString(content)
+  map.setIfUndefined(hashesOut, astHash, (): Ast[] => []).unshift(ast)
+  return astHash
+}
+
+function hashTree(root: Ast) {
+  const hashes = new Map<string, Ast[]>()
+  const rootHash = hashSubtree(root, hashes)
+  return { root: rootHash, hashes }
+}
+
+function rawBlockToInline(tree: RawAst.Tree.Tree) {
+  if (tree.type !== RawAst.Tree.Type.BodyBlock) return tree
+  const statements = Array.from(tree.statements)
+  if (statements.length !== 1) return tree
+  const expression = statements[0]!.expression
+  if (expression === undefined) return tree
+  return expression
+}
+
+function analyzeCode(code: string, asBlock: boolean, module: MutableModule) {
+  const rawParsedBlock = parseEnso(code)
+  const rawParsed = asBlock ? rawParsedBlock : rawBlockToInline(rawParsedBlock)
+  // NOTE: We could do the initial analysis much more efficiently by operating on the raw tree; we just need to separate
+  // the logic for mapping raw types to `Ast` from the actual construction of the `Ast` types.
```

Comment:
It's an optimization I have in mind, but this is not currently the performance bottleneck anyway, so I'll just delete the comment.

The idea is: The only type differences between the Rust parser and the abstracted Ast are because I've made some finer distinctions in Ast that should actually be pushed into the parser. Once we do that, we can content-hash raw or abstracted trees equivalently, and then we would only need to `abstract` subtrees that are not preserved from the original, avoiding most of the JS allocations.